d3b_utils/s3_contents.py

```diff
@@ -0,0 +1,182 @@
+import csv
+import os
+import re
+from concurrent.futures import ThreadPoolExecutor, as_completed
+from itertools import chain
+
+import boto3
+
+
+def fetch_obj_list_info(pathlist, profile="saml", all_versions=False):
+    """List the metadata for a given set of S3 objects. Like
+    fetch_bucket_obj_info but for a list of file paths instead of a bucket.
+
+    :param pathlist: list of s3 file paths
+    :type pathlist: list
+    :param profile: see fetch_bucket_obj_info
+    :param all_versions: see fetch_bucket_obj_info
+
+    :return: see fetch_bucket_obj_info
+    """
+    bucket_paths = {}
+    for path in pathlist:
+        bucket, key = re.sub(r"^s3://", "", path, count=1).split("/", 1)
+        bucket_paths.setdefault(bucket, set()).add(key)
+
+    def scrape(bucket, prefix):
+        return fetch_bucket_obj_info(
+            bucket,
+            search_prefixes=[prefix],
+            profile=profile,
+            all_versions=all_versions,
+        )
```

Comment:
If we are not interested in empty objects, instead of calling `fetch_bucket_obj_info`, have you considered something like: 

```python
def get_obj_info(bucket, prefix):
    s3 = boto3.resource("s3")
    object = s3.Object(bucket, prefix)
    try:
        object.load()
    except botocore.exceptions.ClientError as e:
        if e.response["Error"]["Code"] == "404":
            print("requested object doesn't exist")
        else:
            print(e.response["Error"]["Code"])
    else:
        return {
            "Key": object.key,
            "LastModified": object.last_modified,
            "ETag": object.e_tag,
            "Size": object.content_length,
            "StorageClass": object.storage_class,
        }
```

this way , if we have a file name that is _also_ a prefix, it only returns the object requested and not objects with the same prefix

Reply:
Requesting the object for every file results in a _lot_ of API requests.

Reply:
how about filtering what is returned by fetch_list_obj_info to only returned the objects that correspond to the requested s3 paths?

Reply:
I don't understand that question.

Reply:
But maybe I misunderstood your first question. Instead of paginating each common prefix you want to fetch it first...and then what?

Reply:
> I don't understand why this would be a good thing. What if you want to put both test.csv and also test.csv/yuck.txt into your list of files to retrieve?

if you want both  test.csv and also test.csv/yuck.txt, you should ask for both.... e.g. 

That said, I think it's already working like that and no change is needed.

```python
>>> print([
...      o["Key"] for o in
...      fetch_obj_list_info(
...          ['s3://kf-study-us-east-1-prd-sd-me0wme0w/test.csv'], profile = None
...      )
... ])
['test.csv']
>>> print([
...      o["Key"] for o in
...      fetch_obj_list_info(
...          ['s3://kf-study-us-east-1-prd-sd-me0wme0w/test.csv', 's3://kf-study-us-east-1-prd-sd-me0wme0w/test.csv.yuck'], profile = None
...      )
... ])
['test.csv', 'test.csv.yuck']
```